src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/object/ObjectKeysFunctionIterator.java

```diff
@@ -118,15 +135,35 @@ public void setNextResult() {
     }
 
     @Override
-    public JavaRDD<Item> getRDDAux(DynamicContext dynamicContext) {
-        _currentDynamicContext = dynamicContext;
-        JavaRDD<Item> childRDD = _iterator.getRDD(dynamicContext);
+    public JavaRDD<Item> getRDDAux(DynamicContext context) {
+        _currentDynamicContext = context;
+        JavaRDD<Item> childRDD = _iterator.getRDD(context);
         FlatMapFunction<Item, Item> transformation = new ObjectKeysClosure();
         return childRDD.flatMap(transformation).distinct();
     }
 
     @Override
-    public boolean initIsRDD() {
-        return _iterator.isRDD();
+    public Dataset<Row> getDataFrame(DynamicContext context) {
+        _currentDynamicContext = context;
```

Comment:
That's a very good place to start.

Now, thinking further, I think that getting the keys of a DataFrame is much faster than getting those inside an RDD, because (i) they can directly be read from the schema into a local list and (ii) and it will always be a small list because this is homogeneous data (so outputting as a DF is always an overkill).

How about having the keys simply be output via the local iterator (similar to count()) if the child iterator is a DF?

I.e.:

- isRDD() and isDF() would both return false if the child iterator is a DF
- in the local iterator, test for _iterator.isDF(), and if so, collect the keys from the schema locally to a local list, and return them one by one (i.e., a big switch in setNextResult() for the local child case and the DF child case).

That way, no heavy machinery is needed to populate a new (and small) DF.

Does it make sense?

getRDDAux() can and should of course be left unchanged, because if the input is an RDD, we may have a very heterogeneous sequence with as many keys as objects :)

This discussion is specific to DataFrames.